In [1]:
!pip install dnspython
!pip install pymongo[srv]
import nltk
import pymongo
import pandas as pd
import numpy as np
import dns
import warnings
warnings.simplefilter("ignore")

     |████████████████████████████████| 194kB 4.7MB/s 


Here I have used amazon video games review data which I got from <a href="http://jmcauley.ucsd.edu/data/amazon/">here</a> <br>
Here data is in json format, I have imported this data into mongodb atlas cloud database just for learning purpose you can directly read json file in pandas 

## Connect with Cloud mongodb atlas database and see available instance of database

In [0]:
client = pymongo.MongoClient("mongodb+srv://user:user@cluster0-rgvhf.gcp.mongodb.net/test?retryWrites=true&w=majority")
#for local mongodb you should do MongoClient("localhost",27017)

In [54]:
client.database_names()

['reviews', 'admin', 'local']

In [4]:
db=client['reviews']
db.collection_names()

['videogames']

In [0]:
collection=db['videogames']

In [6]:
temp=collection.find_one()
print(temp)

{'_id': ObjectId('5e53e995a7c489e13c653798'), 'overall': 2.0, 'verified': True, 'reviewTime': '02 20, 2015', 'reviewerID': 'A2204E1TH211HT', 'asin': '0700026657', 'reviewerName': 'Grandma KR', 'reviewText': 'found the game a bit too complicated, not what I expected after having played 1602, 1503, and 1701', 'summary': 'Two Stars', 'unixReviewTime': 1424390400}


# DATA PREPROCESSING

There is no direct method for see which fields have NaN values<br>
So We are gonna retrieve total number of entries and number of times key is present in data  <br>
Data will be retrieved as list of dictionaries from collection

In [0]:
def do_count(collection):
    temp=collection.find()
    print("\ncollection contains {0} entry".format(temp.count())) #to see length of data

    count_key={}
    for item in temp:
        for key in item:
            if key in count_key:
                count_key[key]+=1
            else:
                count_key[key]=1

    print(count_key)        

In [8]:
do_count(collection)


collection contains 497577 entry
{'_id': 497577, 'overall': 497577, 'verified': 497577, 'reviewTime': 497577, 'reviewerID': 497577, 'asin': 497577, 'reviewerName': 497501, 'reviewText': 497419, 'summary': 497468, 'unixReviewTime': 497577, 'vote': 107793, 'style': 289237, 'image': 3634}


You can see that some fields does not contain value for each entry, we can say that they have NaN value in that entry 

### Now let's retrive all data present in collection 


In [9]:
temp_data=collection.find()
print(temp_data)

print("\n",temp_data[10000])  #you can retrive single entry like this


 {'_id': ObjectId('5e53e99ca7c489e13c655ea8'), 'overall': 2.0, 'verified': False, 'reviewTime': '07 25, 2003', 'reviewerID': 'AGOHW680WF4KR', 'asin': 'B00002SU0O', 'reviewerName': 'Roy Levins', 'reviewText': 'THE GAME IS SLOW AND THERE IS NO WRESTLING MOVES.BUT YOU GET OUTSIDE THE RING ACTION WITH ANY WRESTLER.PS IF YOU WANT WRESTLING GET WWF RAW OR THE ROYAL RUMBLE FOR SNES.', 'summary': 'A SORRY GAME FOR WRESTLING.', 'unixReviewTime': 1059091200}


### Let's make dataframe of data which we will use

In [0]:
def retrive_usefuldata(all_data,reviews,ratings,summary):
    review_list=[]
    rating_list=[]
    summary_list=[]
    for item in all_data:
        if reviews in item:                                       #remove elements which do not contain this field
            if summary in item:                                    #remove elements which do not contain this field
                review=item[reviews]
                rating=item[ratings]
                summary_t=item[summary]
                
                review_list.append(review)
                rating_list.append(rating)
                summary_list.append(summary_t)
    
    dictionary = {'user_reviews': review_list,'review_summary': summary_list ,'user_ratings': rating_list}
        
    useful_data=pd.DataFrame(dictionary)
    return useful_data

Above function will return reviews and corresponding summary and rating in dataframe object, you need to give just key names of this fields as per in dictionary  

In [11]:
data=retrive_usefuldata(temp_data,'reviewText','overall','summary')
data.head()

user_reviews  ... user_ratings
0  found the game a bit too complicated, not what...  ...          2.0
1  I played it a while but it was alright. The st...  ...          4.0
2  I bought this game thinking it would be pretty...  ...          5.0
3                                           ok game.  ...          3.0
4  I liked it and had fun with it, played for a w...  ...          4.0

[5 rows x 3 columns]

now we can perform nltk methods on above dataframe

In [12]:
print(len(data))
print(data.isnull().sum())

497316
user_reviews      0
review_summary    0
user_ratings      0
dtype: int64


As you know before we had 497577 raw and now we have 497316 raw , that's because we had some NaN values which we removed inside funtion <br> so now we have cleaned data for our model

In [13]:
print(dict(data['user_ratings'].value_counts()))

{5.0: 299541, 4.0: 93636, 3.0: 49138, 1.0: 30872, 2.0: 24129}


As you can see we have very imbalanced data but it's not a surprise since we know that people only buy product in which they are intersted so they are highly likely to give more ratings

### Creating Balanced Dataset

There are mainly three methods for handling imbalanced dataset <br>
<ol><li>Under Sampling</li>
    <li>Over Sampling</li>
    <li>SMOTE(Synthetic Minority Oversampling Technique)</li></ol><br>
In Under Sampling we simply remove observations from the majority class As In Over Sampling we simply increase observations from the minority class by repeating them until both observations are same<br>

SMOTE is a very powerful technique which generates new minority samples. You can learn more about this <a href="https://arxiv.org/pdf/1106.1813.pdf">here</a><br>
And You can perform SMOTE in python by referring <a href="https://imbalanced-learn.readthedocs.io/en/stable/generated/imblearn.over_sampling.SMOTE.html">this</a> <br>

<b>Since We have too much data and it will take too much time for training a model on this data so here we will take ten thousand sample for each class</b>

In [0]:
import random
def creating_balanced_dataset(dataset,column,count):
    temp_dict=dict(dataset[column].value_counts())
    indices=[]
    for item in temp_dict:
        temp_indices=dataset[dataset[column]==item].sample(count).index
        indices=temp_indices.union(indices)
    balanced_dataset=dataset.loc[indices]
    balanced_dataset.reset_index(drop=True,inplace=True)
    return balanced_dataset

In [15]:
balanced_data=creating_balanced_dataset(data,'user_ratings',10000)

print(balanced_data['user_ratings'].value_counts())
print(balanced_data.shape)

3.0    10000
5.0    10000
1.0    10000
4.0    10000
2.0    10000
Name: user_ratings, dtype: int64
(50000, 3)


# FEATURE ENGINEERING

In [16]:
import nltk
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [0]:
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import re
stop_words=set(stopwords.words('english'))

In [18]:
balanced_data.head()

user_reviews  ... user_ratings
0  found the game a bit too complicated, not what...  ...          2.0
1  I liked it and had fun with it, played for a w...  ...          4.0
2  I've bought and played ALL of the ANNO games s...  ...          1.0
3  I had Dirt 2 on Xbox 360 and it was an okay ga...  ...          4.0
4  I still haven't figured this one out. Did ever...  ...          2.0

[5 rows x 3 columns]

Let's combine both review columns 

In [19]:
balanced_data['review']=balanced_data['user_reviews']+balanced_data['review_summary']

balanced_data.drop(columns=['user_reviews','review_summary'],inplace=True)
balanced_data.head()

user_ratings                                             review
0           2.0  found the game a bit too complicated, not what...
1           4.0  I liked it and had fun with it, played for a w...
2           1.0  I've bought and played ALL of the ANNO games s...
3           4.0  I had Dirt 2 on Xbox 360 and it was an okay ga...
4           2.0  I still haven't figured this one out. Did ever...

Now let's remove all things except characters <br>
Also remove stop words

In [0]:
def clean_review(data,column):
    data[column] = data[column].apply(lambda x:re.sub('[^a-zA-Z ]+','',str(x)))
    
    data[column] = data[column].apply(nltk.word_tokenize)
    
    data[column] = data[column].apply(lambda x:[word.lower() for word in x])   #transform words in lowercase

    data[column] = data[column].apply(lambda x: [word for word in x if word not in stop_words]) #removing stop words 

    data[column] = data[column].apply(lambda x:' '.join(x))    #join all words as sentenance
    return data

In [21]:
cleaned_data=clean_review(balanced_data,'review')
cleaned_data.head()

user_ratings                                             review
0           2.0  found game bit complicated expected played two...
1           4.0  liked fun played got moneys worth certainly go...
2           1.0  ive bought played anno games since spent hours...
3           4.0  dirt xbox okay game started playing games lapt...
4           2.0  still havent figured one everything instructed...

Classfying five class is very hard, we can not get high accuracy as we can get for binary classification<br>
So we will make comparison between both situation<br>
For binary classification We will group [1,2] as low and [3,4,5] as high

In [22]:
binary_rating=[]
for item in cleaned_data['user_ratings']:
    if item>=3:
        binary_rating.append('high')
    else:
        binary_rating.append('low')
cleaned_data=pd.concat([cleaned_data,pd.Series(binary_rating,name='binary_rating')],axis=1) 
print(cleaned_data['binary_rating'].value_counts(),'\n')
cleaned_data.head()

high    30000
low     20000
Name: binary_rating, dtype: int64 



user_ratings  ... binary_rating
0           2.0  ...           low
1           4.0  ...          high
2           1.0  ...           low
3           4.0  ...          high
4           2.0  ...           low

[5 rows x 3 columns]

<b>creating balanced dataset for binary classification</b>

In [23]:
binary_balanced_data=creating_balanced_dataset(cleaned_data,'binary_rating',20000)

print(binary_balanced_data.shape)
print(binary_balanced_data['binary_rating'].value_counts())

(40000, 3)
high    20000
low     20000
Name: binary_rating, dtype: int64


Now We will perform Tfidf vectorization on text content to convert in numbers since computer works best with number

In [24]:
binary_tfidf = TfidfVectorizer(max_features=20000)
binary_word_matrix=binary_tfidf.fit_transform(binary_balanced_data['review'])
print(binary_word_matrix.shape)

tfidf = TfidfVectorizer(max_features=20000)
word_matrix=tfidf.fit_transform(cleaned_data['review'])
print(word_matrix.shape)

(40000, 20000)
(50000, 20000)


In [0]:
binary_target=binary_balanced_data['binary_rating']
target=cleaned_data['user_ratings']

For training a model in keras we should encode and transform our target variable as follow 

In [26]:
from keras.utils import to_categorical
from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()
encoded=encoder.fit_transform(binary_target)
for_keras = to_categorical(encoded,num_classes=2)

Using TensorFlow backend.


In [27]:
for_keras

array([[0., 1.],
       [1., 0.],
       [0., 1.],
       ...,
       [0., 1.],
       [1., 0.],
       [0., 1.]], dtype=float32)

# MODEL DEVELOPEMENT

In [0]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score 
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix 

We will create three different training and testing set where first one is for binary classification, second one is for multiclass classification and third one is for keras binary classification<br>
I will not train keras model for multiclass classification since it doesn't improve accuracy in classifying as I observe in binary classification

In [0]:
btrain_x,btest_x,btrain_y,btest_y=train_test_split(binary_word_matrix,binary_target,test_size=0.05, random_state=42) 
train_x,test_x,train_y,test_y=train_test_split(word_matrix,target,test_size=0.05, random_state=42)

kbtrain_x,kbtest_x,kbtrain_y,kbtest_y=train_test_split(binary_word_matrix,for_keras,test_size=0.05, random_state=42) 

#### Binary Classification 

In [30]:
binary_lr=LogisticRegression()

binary_lr.fit(btrain_x,btrain_y)

blr_pred=binary_lr.predict(btest_x)

print(accuracy_score(btest_y,blr_pred),'\n')
print(confusion_matrix(btest_y,blr_pred),'\n')
print(classification_report(btest_y,blr_pred),'\n')

print(cross_val_score(binary_lr,btrain_x,btrain_y,cv=10).mean())

0.83 

[[825 172]
 [168 835]] 

              precision    recall  f1-score   support

        high       0.83      0.83      0.83       997
         low       0.83      0.83      0.83      1003

    accuracy                           0.83      2000
   macro avg       0.83      0.83      0.83      2000
weighted avg       0.83      0.83      0.83      2000
 

0.8158947368421053


In [31]:
binary_rf=RandomForestClassifier()

binary_rf.fit(btrain_x,btrain_y)

brf_pred=binary_rf.predict(btest_x)

print(accuracy_score(btest_y,brf_pred),'\n')
print(confusion_matrix(btest_y,brf_pred),'\n')
print(classification_report(btest_y,brf_pred))

0.8085 

[[782 215]
 [168 835]] 

              precision    recall  f1-score   support

        high       0.82      0.78      0.80       997
         low       0.80      0.83      0.81      1003

    accuracy                           0.81      2000
   macro avg       0.81      0.81      0.81      2000
weighted avg       0.81      0.81      0.81      2000



In [32]:
binary_xgb=XGBClassifier()

binary_xgb.fit(btrain_x,btrain_y)

bxgb_pred=binary_xgb.predict(btest_x)

print(accuracy_score(btest_y,bxgb_pred),'\n')
print(confusion_matrix(btest_y,bxgb_pred),'\n')
print(classification_report(btest_y,bxgb_pred))

0.7475 

[[722 275]
 [230 773]] 

              precision    recall  f1-score   support

        high       0.76      0.72      0.74       997
         low       0.74      0.77      0.75      1003

    accuracy                           0.75      2000
   macro avg       0.75      0.75      0.75      2000
weighted avg       0.75      0.75      0.75      2000



In [0]:
from keras.layers import Dense,Dropout
from keras.models import Sequential

In [37]:
binary_model = Sequential()
binary_model.add(Dense(10000, activation='relu', input_dim=20000))
binary_model.add(Dropout(0.1))
binary_model.add(Dense(1000, activation='relu'))
binary_model.add(Dropout(0.1))
binary_model.add(Dense(100, activation='relu'))
binary_model.add(Dense(2, activation='softmax'))

binary_model.compile(loss='categorical_crossentropy',optimizer='rmsprop',metrics=['accuracy'])





Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.




In [38]:
binary_model.fit(kbtrain_x,kbtrain_y,epochs=3, batch_size=128)

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Epoch 1/3





38000/38000 [==============================] - 34s 900us/step - loss: 0.4396 - acc: 0.7919
Epoch 2/3
38000/38000 [==============================] - 23s 612us/step - loss: 0.3035 - acc: 0.8716
Epoch 3/3
38000/38000 [==============================] - 23s 611us/step - loss: 0.1746 - acc: 0.9361


In [40]:
keras_pred=binary_model.predict(kbtest_x)

train_score = binary_model.evaluate(kbtrain_x,kbtrain_y)

test_score = binary_model.evaluate(kbtest_x,kbtest_y)

print("training accuracy:{0}".format(train_score[1]),'\n')
print("testing accuracy:{0}".format(test_score[1]),'\n')

print(keras_pred)

2000/2000 [==============================] - 0s 171us/step
training accuracy:0.9745263157894737 

testing accuracy:0.808 

[[0.02259083 0.9774091 ]
 [0.00215436 0.99784565]
 [0.13627647 0.8637235 ]
 ...
 [0.7304234  0.26957664]
 [0.9893373  0.01066263]
 [0.8059728  0.19402719]]


As you can see logistic regression works best and it's also very simple algorithm as it simply draws line between two class<br>
rf and keras also performed well<br>
I would not recommend using SVM because it's very high computational and time consuming algorithm and accuracy is not increasing that much

#### Multiclass Classification

In [33]:
lr=LogisticRegression()

lr.fit(train_x,train_y)

lr_pred=lr.predict(test_x)

print(accuracy_score(test_y,lr_pred),'\n')
print(confusion_matrix(test_y,lr_pred),'\n')
print(classification_report(test_y,lr_pred))

0.5572 

[[356 113  29   9   9]
 [121 213  98  41  19]
 [ 42 104 224 106  39]
 [ 21  33  93 245  95]
 [ 19  11  25  80 355]] 

              precision    recall  f1-score   support

         1.0       0.64      0.69      0.66       516
         2.0       0.45      0.43      0.44       492
         3.0       0.48      0.43      0.46       515
         4.0       0.51      0.50      0.51       487
         5.0       0.69      0.72      0.71       490

    accuracy                           0.56      2500
   macro avg       0.55      0.56      0.55      2500
weighted avg       0.55      0.56      0.55      2500



In [34]:
rf=RandomForestClassifier()

rf.fit(train_x,train_y)

rf_pred=rf.predict(test_x)

print(accuracy_score(test_y,rf_pred),'\n')
print(confusion_matrix(test_y,rf_pred),'\n')
print(classification_report(test_y,rf_pred))

0.5056 

[[375  85  31  14  11]
 [139 188  89  44  32]
 [ 82 111 179  98  45]
 [ 49  43  82 218  95]
 [ 30  35  28  93 304]] 

              precision    recall  f1-score   support

         1.0       0.56      0.73      0.63       516
         2.0       0.41      0.38      0.39       492
         3.0       0.44      0.35      0.39       515
         4.0       0.47      0.45      0.46       487
         5.0       0.62      0.62      0.62       490

    accuracy                           0.51      2500
   macro avg       0.50      0.50      0.50      2500
weighted avg       0.50      0.51      0.50      2500



In [35]:
xgb=XGBClassifier()

xgb.fit(train_x,train_y)

xgb_pred=xgb.predict(test_x)

print(accuracy_score(test_y,xgb_pred),'\n')
print(confusion_matrix(test_y,xgb_pred),'\n')
print(classification_report(test_y,xgb_pred))

0.4568 

[[363  83  14  33  23]
 [156 148  62  71  55]
 [114  81 138 112  70]
 [ 74  47  51 196 119]
 [ 65  21  15  92 297]] 

              precision    recall  f1-score   support

         1.0       0.47      0.70      0.56       516
         2.0       0.39      0.30      0.34       492
         3.0       0.49      0.27      0.35       515
         4.0       0.39      0.40      0.40       487
         5.0       0.53      0.61      0.56       490

    accuracy                           0.46      2500
   macro avg       0.45      0.46      0.44      2500
weighted avg       0.45      0.46      0.44      2500



Here also logistic regression outperformed other algorithms and it's also easy to train 

To increase accuracy you can perform hyperparameter optimization, you can also combine two or three algorithms and create hybrid model which will also perform better <br>
However I would not recommend increasing training sample or feature set because I have already done that and it will not improve accuracy.